src/cripts/README.md

```diff
@@ -0,0 +1,453 @@
+
+## 3.2. Hooks
+
+To simplify usage, Cript predefines a set of function names, each corresponding
+to an ATS hook. These names are set in stone, and will be automatically added
+to the ATS core if provided in the Cript. Those hooks not used will not be
+called, obviously.
+
+The current version of Cript supports the following callbacks / hooks, all of which are optional:
+
+| Callback name      | Hook equivalent                | Description                                        |
+| ------------------ | ------------------------------ | -------------------------------------------------- |
+| do_remap()         | none                           | This is the main entry point for all remap plugins |
+| do_post_remap()    | TS_HTTP_POST_REMAP_HOOK        | Called right after the remap rule is done          |
+| do_send_request()  | TS_HTTP_SEND_REQUEST_HDR_HOOK  | Called before making an origin request             |
+| do_read_response() | TS_HTTP_READ_RESPONSE_HDR_HOOK | Called when the origin has a response              |
+| do_send_response() | TS_HTTP_SEND_RESPONSE_HDR_HOOK | Called before sending a response to client         |
+| do_txn_close()     | TS_HTTP_TXN_CLOSE_HOOK         | Called just before the transaction is done         |
+
+Note that not all of these callbacks are triggered for all requests. For example,
+upon a cache hit in ATS, `do_send_request()` and `do_read_response()` are not called.
+In addition, there are two plugin specific callbacks, which are used when loading
+the plugins and instantiating remap rules:
+
+| Callback name      | API equivalent        | Description                                    |
+| ------------------ | --------------------- | ---------------------------------------------- |
+| do_init            | TSRemapInit()         | Called once when the Cript is loaded           |
+| do_create_instance | TSRemapNewInstance    | Called for every remap rule using the Cript    |
+| do_delete_instance | TSRemapDeleteInstance | Called to cleanup any instance data from above |
+
+## 3.3. Processing and modifying headers
+
+A big part of all Cripts is to read and modify various headers. ATS as it works,
+has four distinct headers, which translates into the following four environments:
+
+| Header name      | Description                                        |
+| ---------------- | -------------------------------------------------- |
+| Client::Request  | The clients request header                         |
+| Client::Response | The response header that is sent to the client     |
+| Server::Request  | The request header being sent to the origin server |
+| Server::Response | The response header being received from origin     |
+
+Accessing these headers is easy:
+```
+    borrow client_req  = Client::Request::get();
+    borrow client_resp = Client::Response::get();
+    borrow server_req  = Server::Request::get();
+    borrow server_resp = Server::Response::get();
+```
```

Comment:
Above we suggest using `auto` liberally. Should we do that in our examples for this, rather than explicitly declare `borrow`?

Reply:
The README.md has not been updated since I first did this, which is why I didn't include it to begin with. But then people wanted something. I'm not going to write a ton of documentation until we decide to actually land and use this :).